library/rsa.c

```diff
@@ -54,6 +54,18 @@
 #include <stdlib.h>
 #endif
 
+/* We use MD first if it's available (for compatibility reasons)
+ * and "fall back" to PSA otherwise (which needs psa_crypto_init()). */
+#if defined(MBEDTLS_PKCS1_V21)
+#if defined(MBEDTLS_MD_C)
+#define HASH_MAX_SIZE   MBEDTLS_MD_MAX_SIZE
+#else /* MBEDTLS_MD_C */
+#include "psa/crypto.h"
```

Comment:
So we're OK with RSA using PSA even if `MBEDTLS_USE_PSA_CRYPTO` is not defined?

Reply:
If `MBEDTLS_PKCS1_V21` is defined then one of `MBEDTLS_MD_C`, `MBEDTLS_PSA_CRYPTO_C` must be defined. Otherwise we have build error:

https://github.com/Mbed-TLS/mbedtls/pull/6141/files#diff-4fc3811c7c53e1746e1056ac38a6462ad2160077a59898bbd2bc31f5938ac185L165

Reply:
Yes, these can be defined, but `MBEDTLS_USE_PSA_CRYPTO` can be undefined to signal that we're not using PSA for (as its doc says) `x509 and TLS library, and enable new APIs for using keys handled by PSA Crypto`. I'm just wondering if it's readable that PSA can be used for something even though `MBEDTLS_USE_PSA_CRYPTO` is undefined.

Reply:
The novelty (use of PSA) is only in builds that where never possible before, so I think it's quite OK from a compatibility perspective.

I don't think we ever meant to promise that nothing would use PSA when `MBEDTLS_USE_PSA` is not defined, but I agree that this is a reasonable way to interpret the current working, so the new behaviour could be surprising to users. I think we can handle that by adding a warning every time something that previously depended on MD now starts depending on "MD or PSA", similar to [this one](https://github.com/Mbed-TLS/mbedtls/pull/6141/files#diff-0279e5915ac49daea5fd9ed14934a074d1a88e9b35c674d7d1a0d0e0688f60bbR1146) - so, eventually X.509 and TLS will get such a warning. And probably we should tune the wording on `MBEDTLS_USE_PSA` as well before the next release to accurately reflect the status as this point - I'll track that as part of #6146.

Reply:
I just thought: anyway, the only case where PSA is used in RSA is when MD is not defined, and the only way you'll be able to compile TLS or X.509 without MD is by enabling `USE_PSA` (they will depend on `MD_C || USE_PSA`). So, X.509 and TLS are indeed only going to use PSA when `USE_PSA` is enabled.